crates/astria-composer/src/searcher/executor.rs

```diff
@@ -164,28 +216,43 @@ impl Executor {
                         }
                     }
+
+                    block_timer.as_mut().reset(Instant::now() + Duration::from_millis(self.block_time_ms));
                 }
 
-                // receive new bundle for processing
-                Some(bundle) = self.new_bundles.recv(), if submission_fut.is_terminated() => {
-                    // TODO(https://github.com/astriaorg/astria/issues/476): Attach the hash of the
-                    // bundle to the span. Linked GH issue is for agreeing on a hash for `SignedTransaction`,
-                    // but both should be addressed.
-                    let span =  info_span!(
-                        "submit bundle",
-                        nonce.initial = nonce,
-                        bundle.len = bundle.len(),
+                Some(bundle) = bundle_factory.pop_complete(), if submission_fut.is_terminated() => {
+                    debug!(
+                        bundle_len=bundle.len(),
+                        "submitting bundle to sequencer"
                     );
-                    submission_fut = SubmitFut {
-                        client: self.sequencer_client.clone(),
-                        address: self.address,
-                        nonce,
-                        signing_key: self.sequencer_key.clone(),
-                        state: SubmitState::NotStarted,
-                        bundle,
+                    submission_fut = self.make_submission_fut(nonce, bundle);
+                }
+
+                // receive new seq_action and bundle it
+                Some(seq_action) = self.seq_actions_rx.recv(), if submission_fut.is_terminated() => {
+                    let seq_action_size = seq_action_size(&seq_action);
+                    match bundle_factory.try_push(seq_action) {
+                        Ok(()) => {}
+                        Err(_seq_action) => {
+                            warn!(
+                                seq_action_size = seq_action_size,
+                                "failed to bundle sequence action: too large. sequence action is dropped."
+                            );
+                        }
+                    }
+                }
+
+                // try to preempt current bundle if the timer has ticked without submitting the next bundle
+                () = &mut block_timer, if submission_fut.is_terminated() => {
+                    if let Some(bundle) = bundle_factory.pop_now() {
+                        debug!(
+                            bundle_len=bundle.len(),
+                            "forcing bundle submission to sequencer"
+                        );
+                        submission_fut = self.make_submission_fut(nonce, bundle);
                     }
-                    .instrument(span)
-                    .fuse();
+
+                    block_timer.as_mut().reset(Instant::now() + Duration::from_millis(self.block_time_ms));
```

Comment:
I don't think this reset here is useful. You have two cases here:

1. the submission terminates in less than the block time => the arm matching on submission completion will reset this
2. the submission exceeds the block time => this arm is disabled until it does, but the timer will be reset either way

Reply:
good point